### PR TITLE
chore(cli): build env deploy into the cli

### PR DESCRIPTION
**Pull request by Lou1415926**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/deploy/env.go [hunk 1]
```diff
@@ -51,6 +54,37 @@ type envDeployer struct {
 	uploadCustomResourceFlag bool
 }
 
+type NewEnvDeployerInput struct {
+	App             *config.Application
+	Env             *config.Environment
+	SessionProvider *sessions.Provider
+}
+
+func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
```
Comment (paragbhingre) [inline internal/pkg/cli/deploy/env.go:68]: Can we add a comment to this public function? Also, where are we calling this function from?

File: internal/pkg/cli/env_deploy.go [hunk 2]
```diff
@@ -42,6 +50,50 @@ type deployEnvOpts struct {
...
+		return nil, err
+	}
+	store := config.NewSSMStore(identity.New(defaultSess), ssm.New(defaultSess), aws.StringValue(defaultSess.Config.Region))
+	ws, err := workspace.New()
+	if err != nil {
+		return nil, fmt.Errorf("new workspace: %w", err)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("initiate a new environment deployer: %w", err)
+	}
```
Comment (paragbhingre) [inline internal/pkg/cli/env_deploy.go]: How does second if will execute if there is an error while creating a new workspace?
And I think second if looks incorrect as it says _initiate a new environment deployer_ which doesn't seem like happening around this code and looks like happening down on line 79

File: internal/pkg/cli/deploy/env.go (hunk 1 above)
Comment (Lou1415926) [reply internal/pkg/cli/deploy/env.go:68]: https://github.com/aws/copilot-cli/blob/0b94c7f4a57b4dd5724331352626c7908553fb0b/internal/pkg/cli/env_deploy.go#L88 Here 😄

File: internal/pkg/cli/env_deploy.go (hunk 2 above)
Comment (Lou1415926) [reply internal/pkg/cli/env_deploy.go]: I forgot to remove the second error handling 😄

File: internal/pkg/deploy/cloudformation/stack/env.go [hunk 3]
```diff
@@ -78,6 +78,7 @@ func (e *EnvStackConfig) Template() (string, error) {
 		return "", err
 	}
 
+	// TODO: remove all these after we are able to migrate to the new upload workflow.
```
Comment (dannyrandall) [inline internal/pkg/deploy/cloudformation/stack/env.go]:
```suggestion
	// TODO(Lou1415926): remove all these after we are able to migrate to the new upload workflow.
```
😜

File: internal/pkg/deploy/cloudformation/stack/env.go [hunk 4]
```diff
@@ -234,12 +235,11 @@ func (e *EnvStackConfig) internalALBSubnets() []string {
...
 func (e *EnvStackConfig) Parameters() ([]*cloudformation.Parameter, error) {
 	httpsListener := "false"
-	if len(e.in.ImportCertARNs) != 0 || e.in.App.Domain != "" {
+	if len(e.importPublicCertARNs()) != 0 || e.in.App.Domain != "" {
 		httpsListener = "true"
 	}
 	internalHTTPSListener := "false"
-	if len(e.in.ImportCertARNs) != 0 && e.in.ImportVPCConfig != nil &&
-		len(e.in.ImportVPCConfig.PublicSubnetIDs) == 0 {
+	if len(e.importPrivateCertARNs()) != 0 {
```
Comment (dannyrandall) [inline internal/pkg/deploy/cloudformation/stack/env.go:242]: yay! 🎉

File: internal/pkg/deploy/cloudformation/stack/env_test.go [hunk 5]
```diff
@@ -264,7 +264,7 @@ func TestEnv_Parameters(t *testing.T) {
 				},
 				{
 					ParameterKey:   aws.String(envParamCreateHTTPSListenerKey),
-					ParameterValue: aws.String("true"),
+					ParameterValue: aws.String("false"),
```
Comment (dannyrandall) [inline internal/pkg/deploy/cloudformation/stack/env_test.go:267]: how was this passing before🤔

**Review by dannyrandall — APPROVED** [review]

File: internal/pkg/cli/deploy/deploy.go [hunk 6]
```diff
@@ -121,12 +121,6 @@ type publicCIDRBlocksGetter interface {
 	PublicCIDRBlocks() ([]string, error)
 }
 
-type customResourcesUploader interface {
```
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/deploy.go:124]: Is this file only related to workload deployment? If so maybe we can think about renaming the file

File: internal/pkg/deploy/cloudformation/stack/env_test.go (hunk 5 above)
Comment (Lou1415926) [reply internal/pkg/deploy/cloudformation/stack/env_test.go:267]: it was related to the change here: https://github.com/aws/copilot-cli/blob/e86f180596c1210f1ae2bb79d44c0e8a4844278e/internal/pkg/deploy/cloudformation/stack/env.go#L238

which was previously:
```go
if len(e.in.ImportCertARNs) != 0 || e.in.App.Domain != ""
```

So when you have imported certs, and 0 public subnet (i.e. imported private cert), this parameter still renders a value of `true` - which, I think, but not sure of, was supposed to be false.

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/deploy/cloudformation/stack/env_test.go (hunk 5 above)
Comment (dannyrandall) [reply internal/pkg/deploy/cloudformation/stack/env_test.go:267]: yeah, i agree it should be false - i'm just confused why it wasn't false _before_ also since [`mockDeployEnvironmentInput`](https://github.com/Lou1415926/copilot-cli/blob/e86f180596c1210f1ae2bb79d44c0e8a4844278e/internal/pkg/deploy/cloudformation/stack/env_test.go#L386) doesn't have any imported certs... anyways. Not a big deal. I'm just glad it's right😊

File: internal/pkg/cli/deploy/deploy.go (hunk 6 above)
Comment (Lou1415926) [reply internal/pkg/cli/deploy/deploy.go:124]: addressed!

File: internal/pkg/cli/env_deploy.go (hunk 2 above)
Comment (Lou1415926) [reply internal/pkg/cli/env_deploy.go]: removed and addressed!
